core/src/main/java/com/linecorp/armeria/internal/server/JacksonUtil.java

```diff
@@ -20,43 +20,65 @@
 
 import java.util.List;
 import java.util.ServiceLoader;
+import java.util.Set;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
-import com.fasterxml.jackson.databind.Module;
+import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.json.JsonMapper;
-import com.google.common.collect.Iterables;
+import com.google.common.collect.ImmutableList;
 
-import com.linecorp.armeria.common.JacksonModuleProvider;
+import com.linecorp.armeria.common.JacksonObjectMapperProvider;
 
 public final class JacksonUtil {
 
     private static final Logger logger = LoggerFactory.getLogger(JacksonUtil.class);
 
+    private static final List<JacksonObjectMapperProvider> providers =
+            ImmutableList.copyOf(ServiceLoader.load(JacksonObjectMapperProvider.class));
+
     private static boolean noticed;
 
     private static final ObjectMapper INSTANCE = newDefaultObjectMapper();
 
     public static ObjectMapper newDefaultObjectMapper() {
-        final JsonMapper.Builder jsonMapperBuilder = JsonMapper.builder();
-        final ServiceLoader<JacksonModuleProvider> providers = ServiceLoader.load(JacksonModuleProvider.class);
-        if (Iterables.isEmpty(providers)) {
+        if (providers.isEmpty()) {
+            // Create the default ObjectMapper with the modules provided by SPI.
+            final JsonMapper.Builder jsonMapperBuilder = JsonMapper.builder();
```

Comment:
Instead of special-casing this, would it make sense to have a fallback provider instance? i.e.
```
private static final JacksonObjectMapperProvider provider;
static {
  if provider found via SPI {
    provider = the first provider found via SPI
  } else {
    provider = DefaultJacksonObjectMapperProvider.INSTANCE;
  }
}
```